[Elbert Benedict] iP

Elben85 opened the pull request:
### Botto
> Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Botto manages your task for you so you don't have to do it.
Botto is

- Text-based
- Easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is

1. download it from [here](https://github.com/Elben85/ip)
2. double-click it
3. add your tasks
4. let it manage tasks for you

And it is **FREE** 😄 

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

***

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

 ```
 public class Launcher {
     public static void main(String[] args) {
         Application.launch(Main.class, args);
     }
 }
 ```

Elben85 committed:
Level-1

Elben85 committed:
Add Task List

Elben85 committed:
Add mark/unmark as done

Elben85 committed:
Add Todos, events, deadlines

Elben85 committed:
Add Automated Text UI Testing

Elben85 committed:
Add Error Handling

Elben85 committed:
Add Delete Task

Elben85 committed:
Add Save Functionality

Elben85 committed:
Add Date and Time Functionality

Elben85 committed:
Merge pull request #1 from Elben85/branch-Level-7

Add Save Functionality

Elben85 committed:
Clear duke.txt content

Elben85 committed:
Modify Duke.txt

Elben85 committed:
Modify .gitignore

Elben85 committed:
Modify .gitignore 2

Elben85 committed:
Modify .gitignore 3

Elben85 committed:
Untrack data/duke.txt

Elben85 committed:
Add Unfound Save File Handling

Elben85 committed:
Modidy duke.txt temporarily

Elben85 committed:
Merge with origin/master

Elben85 committed:
Untrack data/duke.txt

Elben85 committed:
Merge pull request #2 from Elben85/branch-Level-8

Add Date and Time Functionality

Elben85 committed:
Add More OOP

Elben85 committed:
Merge pull request #3 from Elben85/branch-A-MoreOOP

Add More OOP

Elben85 committed:
Organize into package

Elben85 committed:
Merge pull request #4 from Elben85/branch-A-Packages

Organize into package

Elben85 committed:
Add Gradle Support

Elben85 committed:
Merge pull request #5 from Elben85/branch-A-Gradle

Add Gradle Support

Elben85 committed:
Add JUnit Tests

Elben85 committed:
Merge pull request #6 from Elben85/branch-A-JUnit

Add JUnit Tests

Elben85 committed:
Add Jar File

Elben85 committed:
Merge pull request #7 from Elben85/branch-A-Jar

Add Jar File

Elben85 committed:
Add Checkstyle

Elben85 committed:
Add JavaDocs

Elben85 committed:
Fix Styling Error

Elben85 committed:
Rename Test Methods

Elben85 committed:
Add Find Task Functionality

Elben85 committed:
Merge pull request #8 from Elben85/branch-A-JavaDoc

Add JavaDocs

Elben85 committed:
Resolved conflicts with master

Elben85 committed:
Merge pull request #10 from Elben85/branch-A-CodingStandard

Follow coding standard

Elben85 committed:
Resolve conflicts with master

Elben85 committed:
Merge pull request #9 from Elben85/branch-Level-9

Add Find Task Functionality

Elben85 committed:
Add JavaFX dependency

Elben85 committed:
Add GUI Skeleton

Elben85 committed:
Modify Workflow

Elben85 committed:
Fix Missing Welcome Message

Elben85 committed:
Modify Styling

Elben85 committed:
Remove Ui Class Unit Test

Elben85 committed:
Modify JavaDocs

Elben85 committed:
Fix Code Styling

Elben85 committed:
Merge pull request #11 from Elben85/branch-Level-10

Branch level 10

yuehernkang commented:
Overall, all the coding standard rules are adhered to and has high coverage of JavaDocs for all functions. However, perhaps you could organize the different code files into folders such as Ui, Task, Storage, etc.

yuehernkang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+        String command = splitted[0];
+
+        switch (command) {
+        case "todo":
```
Perhaps you can have a Constants file storing all the strings such as "todo", "deadline" and "event"

yuehernkang commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,80 @@
+
+import org.junit.jupiter.api.Test;
+
+public class DeadlineTest {
```
Very well tested as many scenarios are tested.

yuehernkang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+        switch (command) {
+        case "todo":
+            //No Description Given
+            if (splitted.length < 2) {
```
I think it is very good that User Errors such as not entering description are caught!

yuehernkang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,87 @@
+                boolean isDone;
+
+                switch (taskType) {
+                case "[T]":
```
[D], [T], [E], etc. are commonly used. I feel they could be constants? In case next time you might wanna change them, you only have to change at one place instead of searching the whole project for usage of these strings.

yuehernkang commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+    public String delete(int index) throws DukeException {
+        //Index out of bounds
+        if (index > taskList.size() || index < 1) {
+            throw new DukeException("Index Is Not Valid");
```
"Index Is Not Valid" can be saved as a constant too as it is used multiple times.

yuehernkang reviewed (commented):
Just some general comments about cosmetic changes to code.

AshiqurRah commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    public static String[] parseUserInput(String input) throws DukeException {
+        String[] parsed;
+        //Split by regex to get command
```
I like that you included comments that made your code even easier to read. Good job!

AshiqurRah commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+     * @throws DukeException If input is not valid.
+     */
+    public static String[] parseUserInput(String input) throws DukeException {
+        String[] parsed;
```
The name for String array, parsed, sounds ambiguous. A suggestion for the name would be parsedList

AshiqurRah reviewed (commented):
Looks good. Code is easily readable

Elben85 committed:
Add Assertions

There are several methods which assumes that the parameters fulfill a certain conditions and will fail otherwise.

Using assertions help to check whether the assumptions is fulfilled by the program during runtime.

Let's add important assertions that should holds during the program runtime.

Elben85 committed:
Improve Code Quality

The current code has a lot of reused code and method which are too long.

Improving code quality can improve the readability of the code and make it easier to understand..

Let's refactor the code to improve the code quality.

Elben85 committed:
Merge branch 'master' into branch-A-Assertions

Elben85 committed:
Merge pull request #12 from Elben85/branch-A-Assertions

Branch a assertions

Elben85 committed:
Resolve merge conflict

Elben85 committed:
Merge pull request #13 from Elben85/branch-A-CodeQuality

Improve Code Quality

Elben85 committed:
Add Priority Command

Elben85 committed:
Update JUnit Test

Elben85 committed:
Fix Code Style

Elben85 committed:
Merge pull request #14 from Elben85/branch-BCD-Extension

Add Priority

Elben85 committed:
Fix Test Code Style

Elben85 committed:
Reorganize task package unit tests

Elben85 committed:
Add TaskList Unit Tests

Elben85 committed:
Add More Parser Unit Tests

Elben85 committed:
Fix Code Styling

Elben85 committed:
Merge pull request #16 from Elben85/branch-A-MoreTesting

Branch a more testing

Elben85 committed:
Improve GUI Appearance

Elben85 committed:
Merge pull request #17 from Elben85/branch-A-BetterGUI

Improve GUI Appearance

Elben85 committed:
Add User Guide

Elben85 committed:
Fix JavaDocs

Elben85 committed:
Extract Reused Code

Elben85 committed:
Make Find Command Case-Insensitive

Elben85 committed:
Remove Unused Import

Elben85 committed:
Fix README.md

Elben85 committed:
Merge pull request #18 from Elben85/branch-A-UserGuide

Add user guide

Elben85 committed:
Fix README.md

Elben85 committed:
Modify README.md

Elben85 committed:
Fix UG

Elben85 committed:
Attempt to Fix UG

Elben85 committed:
Add Jekyl Theme

Elben85 committed:
Change Remote Theme

Elben85 committed:
Fix Storage Issue

Elben85 committed:
Make Stage Non-Resizeable

Elben85 committed:
Modify README.md
